dist/bin/scala

```diff
@@ -27,147 +26,211 @@ if [ -z "$PROG_HOME" ] ; then
   cd "$saveddir"
 fi
 
-addJvmOptions () {
-  jvm_options+=("'$1'")
-}
+source "$PROG_HOME/bin/common"
 
-addScalacOptions () {
-  java_options+=("'$1'")
+# This script operates in one of 3 usage modes:
+#   script
+#   repl
+#   run
+# execute_mode replaces mutually exclusive booleans:
+#  execute_repl=false
+#  execute_run=false
+#  execute_script=false
+setExecuteMode () {
+  case "${execute_mode-}" in
+  "") execute_mode="$1" ; shift ;;
+  *) echo "execute_mode==[${execute_mode-}], attempted overwrite by [$1]" 1>&2
+    exit 1
+    ;;
+  esac
 }
 
-source "$PROG_HOME/bin/common"
 
-declare -a residual_args
-declare -a script_args
-execute_repl=false
-execute_run=false
-execute_script=false
-with_compiler=false
-class_path_count=0
-CLASS_PATH=""
-save_compiled=false
+with_compiler=false             # to add compiler jars to repl classpath
+let class_path_count=0 || true  # count classpath args, warning if more than 1
+save_compiled=false             # to save compiled script jar in script directory
+CLASS_PATH="" || true           # scala classpath
 
 # Little hack to check if all arguments are options
 all_params="$*"
 truncated_params="${*#-}"
 # options_indicator != 0 if at least one parameter is not an option
 options_indicator=$(( ${#all_params} - ${#truncated_params} - $# ))
 
-[ -n "$SCALA_OPTS" ] && set -- $SCALA_OPTS "$@"
+[ -n "${SCALA_OPTS-}" ] && set -- $SCALA_OPTS "$@"
 
 while [[ $# -gt 0 ]]; do
   case "$1" in
     -repl)
-      execute_repl=true
+      setExecuteMode 'repl'
       shift
       ;;
     -run)
-      execute_run=true
+      setExecuteMode 'run'
       shift
       ;;
-    -cp | -classpath )
+    -cp | -classpath)
       CLASS_PATH="$2${PSEP}"
-      class_path_count+=1
+      let class_path_count+=1
       shift
       shift
       ;;
-    -cp*|-classpath*)
+    -cp*|-classpath*) # partial fix for #10761
       # hashbang can combine args, e.g. "-classpath 'lib/*'"
       CLASS_PATH="${1#* *}${PSEP}"
-      class_path_count+=1
+      let class_path_count+=1
```

Comment:
Thanks for the explanation above, @philwalk . I'm wondering do we need a `|| true` here and line 74?

Reply:
line 74 is okay because `class_path_count` is only zero on line 50, and always > 0 on line 74.
I did discover some other places that did need to be changed:

reference to residual_args in line 160:
```bash
  if [[ $options_indicator -eq 0 && "${residual_args[@]-}" == "" ]]; then
```

change `"$DEBUG"` to `${DEBUG-}"` in line 227:
```bash
  eval "\"$JAVACMD\"" "${DEBUG-}"  "-classpath \"$repl_cparg\"" "${java_args[@]}" "${residual_args[@]}"
```

Reply:
Thanks for the explanation, @philwalk. Once the two issues you mentioned are addressed, I think we can merge the PR.